src/python/pants/bsp/util_rules/targets.py

```diff
@@ -56,6 +57,38 @@
 _FS = TypeVar("_FS", bound=FieldSet)
 
 
+@union
+@dataclass(frozen=True)
+class BSPResolveFieldFactoryRequest(Generic[_FS]):
+    """Requests an implementation of `BSPResolveFieldFactory` which can filter resolve fields.
+
+    TODO: This is to work around the fact that Field value defaulting cannot have arbitrary
+    subsystem requirements, and so `JvmResolveField` and `PythonResolveField` have methods
+    which compute the true value of the field given a subsytem argument. Consumers need to
+    be type aware, and `@rules` cannot have dynamic requirements.
+
+    See https://github.com/pantsbuild/pants/issues/12934 about potentially allowing unions
+    (including Field registrations) to have `@rule_helper` methods, which would allow the
+    computation of an AsyncFields to directly require a subsystem.
```

Comment:
IMO, the urgency of this feature has bumped up quite a bit. We need the ability for a @union to represent a full interface, _including_ async methods (`@rule_helper`s).